src/hardware/iohandler.cpp

```diff
@@ -27,52 +27,93 @@
 
 //#define ENABLE_PORTLOG
 
-IO_WriteHandler * io_writehandlers[3][IO_MAX];
-IO_ReadHandler * io_readhandlers[3][IO_MAX];
+std::unordered_map<io_port_t, IO_WriteHandler> io_writehandlers[IO_SIZES] = {};
+std::unordered_map<io_port_t, IO_ReadHandler> io_readhandlers[IO_SIZES] = {};
 
-static Bitu IO_ReadBlocked(Bitu /*port*/,Bitu /*iolen*/) {
-	return ~0;
+void port_within_proposed(io_port_t port) {
+	assert(port < std::numeric_limits<io_port_t_proposed>::max());
 }
 
-static void IO_WriteBlocked(Bitu /*port*/,Bitu /*val*/,Bitu /*iolen*/) {
+void val_within_proposed(io_val_t val) {
+	assert(val <= std::numeric_limits<io_val_t_proposed>::max());
 }
 
-static Bitu IO_ReadDefault(Bitu port,Bitu iolen) {
+static io_val_t IO_ReadBlocked(io_port_t /*port*/, Bitu /*iolen*/)
+{
+	return static_cast<io_val_t>(~0);
+}
+
+static void IO_WriteBlocked(io_port_t /*port*/, io_val_t /*val*/, Bitu /*iolen*/)
+{}
+
+static io_val_t IO_ReadDefault(io_port_t port, Bitu iolen);
+void IO_WriteDefault(io_port_t port, io_val_t val, Bitu iolen);
```

Comment:
Good catch; this naming convention of prefixing externally called function with SUBSYSTEM_*, makes sense and helps readability.  (but yeah; I can see how some devs just started slapping in on every function because it looks like a general rule).